archive/imo/imo2006_q5.lean

```diff
@@ -0,0 +1,226 @@
+begin
+  -- The cycle [P(t) - t, P(P(t)) - P(t), ...]
+  let C : cycle ℤ := (periodic_orbit (λ x, P.eval x) t).map (λ x, P.eval x - x),
+  have HC : ∀ {n : ℕ}, (λ x, P.eval x)^[n + 1] t - ((λ x, P.eval x)^[n] t) ∈ C,
+  { intro n,
+    rw [cycle.mem_map, function.iterate_succ_apply'],
+    exact ⟨_, iterate_mem_periodic_orbit ht n, rfl⟩ },
+
+  -- Elements in C are all divisible by one another.
+  have Hdvd : C.chain (∣),
+  { rw [cycle.chain_map, periodic_orbit_chain' _ ht],
+    intro n,
+    convert sub_dvd_eval_sub ((λ x, P.eval x)^[n + 1] t) ((λ x, P.eval x)^[n] t) P;
+    rw function.iterate_succ_apply' },
+
+  -- Any two entries in C have the same absolute value.
+  have Habs : ∀ m n : ℕ, ((λ x, P.eval x)^[m + 1] t - ((λ x, P.eval x)^[m] t)).nat_abs =
+    ((λ x, P.eval x)^[n + 1] t - ((λ x, P.eval x)^[n] t)).nat_abs :=
+  λ m n, nat_abs_eq_of_chain_dvd Hdvd HC HC,
+
+  -- We case on whether the elements on C are pairwise equal.
+  by_cases HC' : C.chain (=),
+  { -- Any two entries in C are equal.
+    have Heq : ∀ m n : ℕ, (λ x, P.eval x)^[m + 1] t - ((λ x, P.eval x)^[m] t) =
+      ((λ x, P.eval x)^[n + 1] t - ((λ x, P.eval x)^[n] t)) :=
+    λ m n, cycle.chain_iff_pairwise.1 HC' _ HC _ HC,
+
+    -- The sign of P^n(t) - t is the same as P(t) - t for positive n. Proven by induction on n.
+    have IH : ∀ n : ℕ, ((λ x, P.eval x)^[n + 1] t - t).sign = (P.eval t - t).sign,
+    { intro n,
+      induction n with n IH,
+      { refl },
+      { apply eq.trans _ (int.sign_add_eq_of_sign_eq IH),
+        have H := Heq n.succ 0,
+        dsimp at H ⊢,
+        rw [←H, sub_add_sub_cancel'] } },
+
+    -- This implies that the sign of P(t) - t is the same as the sign of P^k(t) - t, which is 0.
+    -- Hence P(t) = t and P(P(t)) = P(t).
+    rcases ht with ⟨(_ | k), hk, hk'⟩,
+    { exact (irrefl 0 hk).elim },
+    { have H := IH k,
+      rw [hk'.is_fixed_pt.eq, sub_self, int.sign_zero, eq_comm, int.sign_eq_zero_iff_zero,
+        sub_eq_zero] at H,
+      simp [is_periodic_pt, is_fixed_pt, H] } },
+  { -- We take two nonequal consecutive entries.
+    rw [cycle.chain_map, periodic_orbit_chain' _ ht] at HC',
+    push_neg at HC',
+    cases HC' with n hn,
+
+    -- They must have opposite sign, so that P^{k + 1}(t) - P^k(t) = P^{k + 2}(t) - P^{k + 1}(t).
+    cases int.nat_abs_eq_nat_abs_iff.1 (Habs n n.succ) with hn' hn',
+    { apply (hn _).elim,
+      convert hn';
+      simp only [function.iterate_succ_apply'] },
+
+    -- We deduce P^{k + 2}(t) = P^k(t) and hence P(P(t)) = t.
+    { rw [neg_sub, sub_right_inj] at hn',
+      simp only [function.iterate_succ_apply'] at hn',
+      exact @is_periodic_pt_of_mem_periodic_pts_of_is_periodic_pt_iterate _ _ t 2 n ht hn'.symm } }
+end
+
+@[simp] theorem iterate_comp_eval (P : polynomial ℤ) (k : ℕ) (t : ℤ) :
+  (P.comp^[k] X).eval t = ((λ x, P.eval x)^[k] t) :=
+begin
+  induction k with k IH,
+  { simp },
+  { rw [iterate_succ_apply', iterate_succ_apply', eval_comp, IH] }
+end
```

Comment:
Could you move this one just after `polynomial.eval_comp` in `data.polynomial.basic`, name it `polynomial.eval_comp_iterate`, and generalise `X` to an arbitrary polynomial `Q`? It would be even better to prove the `polynomial.eval₂` statement first.